File: Applied Course/4.Problem Solving/12.Problems on Backtracking/7.Generate Parentheses.py
# https://leetcode.com/problems/generate-parentheses/
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def printAllParenthesis(openStack,closeStack,string):
            if openStack==0 and closeStack==0:
                result.append(string)
            if openStack>0:
                printAllParenthesis(openStack-1,closeStack+1,string+"(")
            if closeStack>0:
                printAllParenthesis(openStack,closeStack-1,string+")")
        result=[]
        if n:
            printAllParenthesis(n,0,"");
        return result

File: Applied Course/4.Problem Solving/12.Problems on Backtracking/7.Generate Parentheses.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<string> result;
    void printAllParenthesis(int openStack,int closeStack,string s)
    {
        if(openStack==0 && closeStack==0)
            result.push_back(s);
        if(openStack>0)
            printAllParenthesis(openStack-1,closeStack+1,s+"(");
        if(closeStack>0)
            printAllParenthesis(openStack,closeStack-1,s+")");
    }
    vector<string> generateParenthesis(int n) {
        if(n)
            printAllParenthesis(n,0,"");
        return result;
    }
};
int main()
{
    // https://leetcode.com/problems/generate-parentheses/
    return 0;
}